src/QsCompiler/Transformations/SyntaxTreeTrimming.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.Quantum.QsCompiler.DependencyAnalysis;
+using Microsoft.Quantum.QsCompiler.SyntaxTree;
+using Microsoft.Quantum.QsCompiler.Transformations.Core;
+
+namespace Microsoft.Quantum.QsCompiler.Transformations.SyntaxTreeTrimming
+{
+    public static class TrimSyntaxTree
+    {
+        public static QsCompilation Apply(QsCompilation compilation, bool keepAllIntrinsics)
+        {
+            return TrimTree.Apply(compilation, keepAllIntrinsics);
+        }
+
+        private class TrimTree : SyntaxTreeTransformation<TrimTree.TransformationState>
+        {
+            public static QsCompilation Apply(QsCompilation compilation, bool keepAllIntrinsics)
+            {
+                var namespaceNames = compilation.Namespaces.Select(ns => ns.Name).ToImmutableHashSet();
+                var augmentedEntryPoints = BuiltIn.AllBuiltIns
+                    .Where(bi => bi.Kind != BuiltInKind.Attribute && namespaceNames.Contains(bi.FullName.Namespace))
+                    .Select(bi => bi.FullName)
+                    .Concat(compilation.EntryPoints)
+                    .Distinct()
+                    .ToImmutableArray();
+                var compilationWithBuiltIns = new QsCompilation(compilation.Namespaces, augmentedEntryPoints);
+                var callablesToKeep = new CallGraph(compilationWithBuiltIns, true).Nodes.Select(node => node.CallableName).ToImmutableHashSet();
+
+                // ToDo: convert to using ternary operator, when target-type
+                // conditional expressions are supported in C#
+                Func<QsNamespaceElement, bool> filter = elem => Filter(elem, callablesToKeep);
+                if (keepAllIntrinsics)
```

Comment:
It would be great to add a comment on why `keepAllIntrinsics` is needed. Maybe in a documentation comment to the `Apply` method.

Is this used when executing this rewrite-step on libraries such that it doesn't remove all callables because they are not called anywhere that starts at an entry-point?

Reply:
Hmm, this behavior was migrated from the Monomorphization rewrite step. Now that Monomorphization doesn't do trimming, it didn't make sense to keep in there. It was originally added by @bettinaheim I think, I'm not entirely sure what it is used for, I had thought maybe it was related to QIR emmission. @bettinaheim could you provide more context for this flag?

Reply:
Yes, I can. The behavior before introducing the QIR setup was to always keep intrinsics as is in the compilation, since the C# runtime relies on abstract classes being generated for all of the intrinsics, independent on whether they are used or not. The new interface based approach for C# emission (which has been implemented but we haven't fully enabled it yet), will remove that requirement I believe (@swernli for confirmation). 
Conversely, the C++ runtime does not rely on any declarations in the emitted llvm code, and it hence makes much more sense to only generate the declarations that are actually needed, such that it is possible to run code using a runtime with limited capabilities (e.g. our bigint support is still missing), as long as the program doesn't actually require anything that is not implemented. Hence, for qir emission, we monomorphize and trim the intrinsics as well.

Reply:
Thanks for the explanation!